crypto/x509/x509_test.cc

```diff
@@ -2301,6 +2523,58 @@ TEST(X509Test, Ed25519Sign) {
   ASSERT_TRUE(SignatureRoundTrips(md_ctx.get(), pub.get()));
 }
 
+TEST(X509Test, Dilithium3SignVerifyCert) {
+  //generate the dilithium key
+  EVP_PKEY_CTX *pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+  ASSERT_NE(pkey_ctx, nullptr);
+  EVP_PKEY *pkey = EVP_PKEY_new();
+  ASSERT_NE(pkey, nullptr);
+  EXPECT_TRUE(EVP_PKEY_keygen_init(pkey_ctx));
+  EXPECT_TRUE(EVP_PKEY_keygen(pkey_ctx, &pkey));
+
+  //generate the cert
+  bssl::UniquePtr<X509> leaf =
+      MakeTestCert("Intermediate", "Leaf", pkey, /*is_ca=*/false);
+  ASSERT_TRUE(leaf);
+
+  //sign the cert
+  bssl::ScopedEVP_MD_CTX md_ctx;
+  EVP_DigestSignInit(md_ctx.get(), nullptr, nullptr, nullptr, pkey);
+  ASSERT_TRUE(X509_sign_ctx(leaf.get(), md_ctx.get()));
+
+  //verify the cert
+  ASSERT_TRUE(X509_verify(leaf.get(), pkey));
+
+  EVP_PKEY_CTX_free(pkey_ctx);
+  EVP_PKEY_free(pkey);
+}
+
+TEST(X509Test, TestDilithium3) {
+  // This test generates a Dilithium3 certificate from the PEM encoding,
+  // extracts the public key, and then verifies the certificate.
+  bssl::UniquePtr<X509> cert(CertFromPEM(kDilithium3Cert));
+  ASSERT_TRUE(cert);
+
+  bssl::UniquePtr<EVP_PKEY> pkey(X509_get_pubkey(cert.get()));
+  ASSERT_TRUE(pkey);
+
+  ASSERT_TRUE(X509_verify(cert.get(), pkey.get()));
```

Comment:
what does the valid certificate consist of? If a random byte in the certificate was changed, should `X509_verify` fail? If so, can we add a test for that?

Reply:
>what does the valid certificate consist of? 

`X509Test, TestDilithium3` is following the suite of tests: [ed25519](https://github.com/jakemas/aws-lc/blob/8442729b7965038b56f2d324e1b6631a36103255/crypto/x509/x509_test.cc#L2082-L2090), [RsassaPss](https://github.com/jakemas/aws-lc/blob/8442729b7965038b56f2d324e1b6631a36103255/crypto/x509/x509_test.cc#L2062-L2069), [RSAPSS](https://github.com/jakemas/aws-lc/blob/8442729b7965038b56f2d324e1b6631a36103255/crypto/x509/x509_test.cc#L2052-L2060) that test the generation of a certificate from PEM, extraction of the public key, and then verification using this public key.

>If a random byte in the certificate was changed, should X509_verify fail? If so, can we add a test for that?

That is what we have added testing for in `X509Test, TestBadDilithium3`. We took a valid certificate, modified the bytes (to make it invalid) and then attempt verification. This is following the similar tests for [PSS](https://github.com/jakemas/aws-lc/blob/8442729b7965038b56f2d324e1b6631a36103255/crypto/x509/x509_test.cc#L2071-L2080), [ed25519](https://github.com/jakemas/aws-lc/blob/8442729b7965038b56f2d324e1b6631a36103255/crypto/x509/x509_test.cc#L2092-L2105).

Reply:
yes, my question is can we add more tests? I'd like to have tests that make sure that `X509_verify` function correctly identifies all different types of invalid certificates and returns appropriate error codes, because letting an invalid certificate pass is a catastrophic bug/security issue. For example, the invalid case you mention in another comment: "modifying the bytes at offset 31 to add NULL parameter selection, as the draft standard for dilithium also proposes that these are absent."

Reply:
New test added in [e9f9c67](https://github.com/awslabs/aws-lc/pull/759/commits/e9f9c6765adb832ac55ff6561022ab244e261726) that checks params are NULL.